source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown

```diff
@@ -0,0 +1,348 @@
+
+Everybody walks in through the front door - thieves and students alike.
+
+I thought about that problem recently in the context of our GraphQL schema. Here we have another open door, this time the controller endpoint that receives and processes the query string. The authentication happens early; the user is admitted without yet knowing what they're after.
+
+Let's say our user makes a query that looks like this:
+
+```
+query {
+  delayedJobs {
+    createdAt
+    failedAt
+    handler
+  }
+}
+```
+
+When our server processes the query (using the [`graphql` gem](https://github.com/rmosolgo/graphql-ruby)), it executes [three steps](https://dev-blog.apollodata.com/graphql-explained-5844742f195e) in the following order:
+
+**1. Parse**
+
+The server constructs an abstract syntax tree (AST) representing the query and validates the query syntax.
+
+**2. Validate**
+
+The server validates the query path, ensuring that the hierarchy of requested fields is correct and that any required arguments are provided, among other checks. A [query analyzer](http://graphql-ruby.org/queries/analysis.html) visits each field *without* executing the field's resolver and collects metadata before returning a final result.
+
+**3. Execute**
+
+The server resolves each field, beginning with the query root and working its way down the AST until it hits a leaf or an `ExecutionError`.
+
+In the example above, assuming our query syntax is properly formed, the server should quickly iterate through all three steps and return an array of `DelayedJob` objects, each with the `createdAt`, `failedAt`, and `handler` properties.
+
+Of course, we've got a problem. It's perfectly reasonable for an administrator to request our `delayedJobs` - however, we don't want to share that information with just *anybody*. In fact, we'd like to prevent the majority of our users from even being able to access this field.
+
+## Ahead-of-Time Authorization
+Incidentally, access-level authorization lines up nicely with the second step in the query processing flow. We only need to know enough about the field to make a quick decision as to whether or not a user should be able to request it - we don't actually care about what the field returns (in fact, we don't even want the resolver to execute).
+
+Let's see if we can use a query analyzer to prevent the field from surfacing:
+
+```ruby
+TestSchema = GraphQL::Schema.define do
+  query_analyzer AuthorizationAnalyzer.new
+end
+```
+
+We'll define our new `AuthorizationAnalyzer` using the template provided in the `graphql` gem documentation:
+
+```ruby
+class AuthorizationAnalyzer
+  # Called before the visit.
+  # Returns the initial value for `memo`
+  def initial_value(query)
+  end
+
+  # This is like the `reduce` callback.
+  # The return value is passed to the next call as `memo`
+  def call(memo, visit_type, irep_node)
+  end
+
+  # Called when we're done the whole visit.
```

Comment:
I think you accidentally a word.

Reply:
Heh, this is a straight copy from the GraphQL Ruby docs: http://graphql-ruby.org/queries/analysis.html. Good catch, though!